changes/2024-05-20-keystore-kms-config/change.md

```diff
@@ -0,0 +1,274 @@
+However, Tenants may want an MPL Consumer to protect
+their data with a particular KMS Key that is NOT necessarily the same
+as other tenants.
+
+Such a Multi-KMS Keystore is, prior to this change, difficult to use
+as a Keystore instance is restricted to one and only one KMS Key ARN.
+Many MPL Consumers have developed a work around to this restriction
+by constructing and then caching a Keystore per KMS Key ARN.
+These MPL Consumers than stand up their own "request-to-Keystore-to-BranchKeyID" logic.
+This workaround may require an additional lookup
+compared to the normal "Encryption Context determines Branch Key ID"
+path.
+
+The goal then is to allow MPL Consumers to construct Keystore instances that
+MAY use the KMS Key ARNs already persisted to a Logical Keystore.
+
+Put another way,
+usage of Branch Keys need only consider the Keystore's Logical Name
+(and therefore, backing persistance medium, which is currently always a DynamoDB table)
+and Branch Key ID.
+
+Usage may not need the KMS ARN.
+
+### KMS Alias
+
+Because the Branch Key creation operation asserted the result
+of KMS `ReEncrypt` matched the Keystore's configured KMS Key ARN,
+a KMS Key Alias could never be used with a Keystore to successfully create a Branch Key.
+
+However, the runtime exception thrown did not make it clear
+why `CreateKey` failed.
+
+Rather than correct the exception,
+Keystore construction MUST fail if the KMS Configuration
+is "strict"
+(KMS Key ARN or KMS MRKey ARN)
+and the provided ARN is an Alias.
+
+### Hierarchy Keyring Caching
+
+Currently, the Hierarchy Keyring create's it's own cache.
+
+To facilitate Caching performance across Keystores/KMS Clients/KMS Keys,
+we could (optionally) break the Cache out of the Hierarchy Keyring.
+
+The Hierarhcy Keyring would instead take a constructed a Cache as
+an arguement during initialization.
+
+When writing to the Cache,
+the Hierarchy Keyring would include the Keystore's ID,
+which, ideally, uniquely identifies it's KMS relationship.
+
+By KMS Relationship, we mean any or all of the following:
+
+- KMS Configuration
+- Credentials used when creating the KMS Client, and thus used when calling KMS
+- Other properties of the KMS Client, such as the region, or request headers
+
+The changes proposed (and comitted to) here
+do not include such a Cache refactoring,
+as that would not address the stated goal
+of allowing one Keystore to work across multiple
+KMS ARNs.
+
+## Operational Implications
+
+For Keystore's usage by a AWS KMS Hierarchical Keyring,
+a "discovery Keystore" is identical to
+a "kmsKeyArn Keystore".
+
+However, only a "strict Keystore" can Create
+or Version (rotate) Branch Keys.
+
+Thus, all MPL Consumers MAY or MAY NOT
+adopt a "discovery Keystore" as best fits their needs.
+
+## Reference-level Explanation
+
+### Keystore [AWS KMS Configuration](../../framework/branch-key-store.md#aws-kms-configuration) Options
+
+The following Options are introduced:
+
+- Discovery
+- MRDiscovery
```

Comment:
Just wondering why MRDiscovery and not MRKDiscovery?

Reply:
Multi-Region Key Discovery sounds silly compared to Multi Region Discovery.
Unless you want to rename both to 
- kmsKeyDiscovery
- kmsMRKeyDiscovery

Reply:
FWIW, I prefer MRKDiscovery. "MR" is never used as an acronym, whereas MRK is.

Reply:
1. MR is clearly used as acronym in the acronym MRK. Just as US is used in USA.
2. If the "key" element needs to be spelled out in MRDiscovery, why do we not include in Discovery?

Your options are:
- Discovery & MRDiscovery
OR
- KeyDiscovery & MRKDiscovery

I am not putting Key in one and not the other.

Reply:
Now that we have launched, this ship has sailed.
Resolving.